File: memory-game.pl
#!/usr/bin/perl

use strict;
use warnings;

# initialize game board
my @cards = (
    ["1", "2", "3", "4"],
    ["5", "6", "7", "8"],
    ["1", "2", "3", "4"],
    ["5", "6", "7", "8"]
);

my @board = (
    [" ", " ", " ", " "],
    [" ", " ", " ", " "],
    [" ", " ", " ", " "],
    [" ", " ", " ", " "]
);

my $num_pairs = 8; # number of pairs of cards to match
my $num_matches = 0; # number of matched pairs so far

# shuffle cards
for (my $i = 0; $i < @cards; $i++) {
    for (my $j = 0; $j < @{$cards[$i]}; $j++) {
        my $x = int(rand(@cards));
        my $y = int(rand(@{$cards[$x]}));
        ($cards[$i][$j], $cards[$x][$y]) = ($cards[$x][$y], $cards[$i][$j]);
    }
}

# main game loop
while ($num_matches < $num_pairs) {
    # print board
    print "   1  2  3  4\n";
    print "  -----------\n";
    for (my $i = 0; $i < @board; $i++) {
        print "$i |";
        for (my $j = 0; $j < @{$board[$i]}; $j++) {
            if ($board[$i][$j] eq "X") {
                print " X ";
            } else {
                print " " . $cards[$i][$j] . " ";
            }
            print "|";
        }
        print "\n";
    }
    print "  -----------\n\n";

    # prompt user for move
    my ($row1, $col1, $row2, $col2);
    do {
        print "Enter the coordinates of the first card to flip (row column): ";
        ($row1, $col1) = split(" ", <STDIN>);
    } until ($board[$row1][$col1] eq " ");
    $board[$row1][$col1] = "X";

    do {
        print "Enter the coordinates of the second card to flip (row column): ";
        ($row2, $col2) = split(" ", <STDIN>);
    } until ($board[$row2][$col2] eq " " && ($row1 != $row2 || $col1 != $col2));
    $board[$row2][$col2] = "X";

    # check for match
    if ($cards[$row1][$col1] eq $cards[$row2][$col2]) {
        print "You found a match!\n";
        $num_matches++;
        $board[$row1][$col1] = " ";
        $board[$row2][$col2] = " ";
    } else {
        print "No match. Try again.\n";
        $board[$row1][$col1] = " ";
        $board[$row2][$col2] = " ";
    }
}

print "Congratulations! You won!\n";

File: memory-game.r
# Define the cards and their corresponding symbols
cards <- c("A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z")
symbols <- sample(cards, length(cards)/2)
symbols <- c(symbols, symbols)
symbols <- sample(symbols)

# Function to print the game board
print_board <- function(board) {
  for (i in 1:nrow(board)) {
    for (j in 1:ncol(board)) {
      if (board[i,j] == "") {
        cat("  ")
      } else {
        cat(board[i,j], " ")
      }
    }
    cat("\n")
  }
}

# Function to check if the game is over
game_over <- function(board) {
  sum(board == "") == 0
}

# Game loop
board <- matrix("", nrow=5, ncol=10)
attempts <- 0
while (TRUE) {
  # Print the game board
  print_board(board)
  
  # Get the user's guess
  cat("\n")
  guess1 <- readline(prompt="Enter the row of your first guess: ")
  guess1 <- as.integer(guess1)
  guess2 <- readline(prompt="Enter the column of your first guess: ")
  guess2 <- as.integer(guess2)
  cat("\n")
  guess3 <- readline(prompt="Enter the row of your second guess: ")
  guess3 <- as.integer(guess3)
  guess4 <- readline(prompt="Enter the column of your second guess: ")
  guess4 <- as.integer(guess4)
  
  # Check the guesses and print the result
  if (board[guess1,guess2] == "" & board[guess3,guess4] == "") {
    if (symbols[guess1*10+guess2] == symbols[guess3*10+guess4]) {
      board[guess1,guess2] <- symbols[guess1*10+guess2]
      board[guess3,guess4] <- symbols[guess3*10+guess4]
      cat("Correct!\n")
    } else {
      cat("Incorrect.\n")
    }
  } else {
    cat("Invalid guess.\n")
  }
  
  # Increment the number of attempts
  attempts <- attempts + 1
  
  # Check if the game is over
  if (game_over(board)) {
    cat("\n")
    print_board(board)
    cat("Congratulations! You completed the game in ", attempts, " attempts.\n")
    break
  }
}
